File: IT/2/Multithreading/Esercizi/Java/Esercizio29.java
/* file Esercizio29.java */

import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.ArrayBlockingQueue;

// la "sala d'aspetto" del barbiere contiene tutte le variabili e i semafori per coordinare il lavoro dei thread 
public class Esercizio29 {
	private final int TOT_SEATS = 5;
	private final BlockingQueue <String>customers = new ArrayBlockingQueue<>(TOT_SEATS);

	class Barber implements Runnable {
		@Override
		public void run() {
			Thread t = Thread.currentThread();
			String tb = t.getName();
			while (true) {
				try {
					String tc = customers.take();
					System.out.println("Il barbiere " + tb + " sta tagliando i capelli al cliente " + tc + ".");
					int pause = ThreadLocalRandom.current().nextInt(10, 25+1);
					t.sleep(pause*1000);

					System.out.println(tb + " ha impiegato " + pause + " minuti per tagliare i capelli a " + tc + ". Avanti un altro!");
				}
				catch (InterruptedException e) {
					e.printStackTrace();
				}
			}
		}
	}

	// 
	class Customer implements Runnable {
		@Override
		public void run() {
			String tc = Thread.currentThread().getName();
			System.out.println("Il cliente " + tc + " entra nel negozio.");
			if (customers.remainingCapacity() > 0) {
				customers.add(tc);
				System.out.println(tc + " si siede e aspetta di farsi tagliare i capelli...");
			}
			else {
				System.out.println("Il cliente " + tc + " non ha trovato posto per sedersi ed esce dal negozio...");
			}
		}
	}

	/* il costruttore avvia l'applicazione */

	public Esercizio29() {
		Thread barber = new Thread(new Barber());
		barber.start();

		while (true) {
			try {
				Thread customer = new Thread(new Customer());
				customer.start();
				int pause = ThreadLocalRandom.current().nextInt(5, 30+1);
				Thread.currentThread().sleep(pause*1000);
			}
			catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
	}

	public static void main(String [] args) {
		new Esercizio29();
	}
}

File: IT/2/Multithreading/Esercizi/Python/esercizio29.py
# file esercizio29.py

from __future__ import print_function
from multiprocessing import Process, current_process, Semaphore, Value, Queue
from random import randint
from time import sleep

def barber(customers):
	p = current_process().name
	while True:
		pc = customers.get()

		print("Il barbiere {} sta tagliando i capelli al cliente {}.".format(p, pc))
		pause = randint(10, 25)
		sleep(pause)
		print("{} ha impiegato {} minuti per tagliare i capelli a {}. Avanti un altro!".format(p, pause, pc))

def customer(customers):
	p = current_process().name
	print("Il cliente {} entra nel negozio.".format(p))

	if customers.qsize() < 5:
		customers.put(p)
		print("{} si siede e aspetta di farsi tagliare i capelli...".format(p))
	else:
		print("Il cliente {} non ha trovato posto per sedersi ed esce dal negozio...".format(p))

if __name__ == "__main__":
	# coda sincronizzata tra i processi 
	customers = Queue(maxsize=5)

	pbarber = Process(target=barber, args=(customers,))

	pbarber.start()

	procs = []

	while True:
		pcustomer = Process(target=customer, args=(customers,))
		pcustomer.start()
		procs.append(pcustomer)
		pause = randint(5, 30)
		sleep(pause)

	pbarber.join()

	for c in procs:
		c.join()

